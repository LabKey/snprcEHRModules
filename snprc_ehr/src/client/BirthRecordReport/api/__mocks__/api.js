/* eslint-disable no-unused-vars, camelcase */

import { NewAnimalData } from '../../tests/fixtures/apiTestData'

export const request = ({ schemaName, queryName, viewName = '', sort = '', columns = [], filterArray = [] }) => {
    // console.log(`loading list: ${queryName}`);
    return new Promise(resolve => {
        switch (queryName) {
        case 'NewAnimalData':
            resolve(NewAnimalData)
            break
        default:
            break
        }
    })
}
