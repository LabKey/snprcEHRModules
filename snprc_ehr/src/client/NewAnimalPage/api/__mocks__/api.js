/* eslint-disable no-unused-vars, camelcase */

import { CurrentSpeciesLookup, AccountLookup, validInstitutions, ValidDiet, AcquisitionTypesLookup,
    valid_bd_status, PotentialDams, PotentialSires, ActiveLocationsAll, colonyGroups,
    ProtocolLookup, pedigreeGroups } from '../../tests/fixtures/apiTestData'

export const request = ({ schemaName, queryName, viewName = '', sort = '', columns = [], filterArray = [] }) => {
    // console.log(`loading list: ${queryName}`);
    return new Promise(resolve => {
        switch (queryName) {
        case 'CurrentSpeciesLookup':
            resolve(CurrentSpeciesLookup)
            break
        case 'AccountLookup':
            resolve(AccountLookup)
            break
        case 'validInstitutions':
            resolve(validInstitutions)
            break
        case 'ValidDiet':
            resolve(ValidDiet)
            break
        case 'AcquisitionTypesLookup':
            resolve(AcquisitionTypesLookup)
            break
        case 'PotentialDams':
            resolve(PotentialDams)
            break
        case 'PotentialSires':
            resolve(PotentialSires)
            break
        case 'ActiveLocationsAll':
            resolve(ActiveLocationsAll)
            break
        case 'colonyGroups':
            resolve(colonyGroups)
            break
        case 'ProtocolLookup':
            resolve(ProtocolLookup)
            break
        case 'pedigreeGroups':
            resolve(pedigreeGroups)
            break
        case 'valid_bd_status':
            resolve(valid_bd_status)
            break
        default:
            break
        }
    })
}
