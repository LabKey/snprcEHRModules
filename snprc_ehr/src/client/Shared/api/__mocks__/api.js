/* eslint-disable no-unused-vars, camelcase */

import { NewAnimalData } from '../../../BirthRecordReport/tests/fixtures/apiTestData'
import {
    CurrentSpeciesLookup, AccountLookup, validInstitutions, ValidDiet, AcquisitionTypesLookup,
    valid_bd_status, PotentialDams, PotentialSires, ActiveLocationsAll, colonyGroups,
    ProtocolLookup, pedigreeGroups
} from '../../../NewAnimalPage/tests/fixtures/apiTestData'

export const request = ({ schemaName, queryName, viewName = '', sort = '', columns = [], filterArray = [] }) => {
    // console.log(`loading list: ${queryName}`);
    return new Promise(resolve => {
        switch (queryName) {
            case 'CurrentSpeciesLookup':
                resolve(CurrentSpeciesLookup)
                break
            case 'AccountLookup':
                resolve(AccountLookup)
                break
            case 'validInstitutions':
                resolve(validInstitutions)
                break
            case 'ValidDiet':
                resolve(ValidDiet)
                break
            case 'AcquisitionTypesLookup':
                resolve(AcquisitionTypesLookup)
                break
            case 'PotentialDams':
                resolve(PotentialDams)
                break
            case 'PotentialSires':
                resolve(PotentialSires)
                break
            case 'colonyGroups':
                resolve(colonyGroups)
                break
            case 'ProtocolLookup':
                resolve(ProtocolLookup)
                break
            case 'pedigreeGroups':
                resolve(pedigreeGroups)
                break
            case 'valid_bd_status':
                resolve(valid_bd_status)
                break
            case 'NewAnimalData':
                resolve(NewAnimalData)
                break
            default:
                break
        }
    })
}

export const executeSql = ({ schemaName, sql, sort = '' }) => {
    return new Promise(resolve => {
        if (sql.indexOf('FROM ehr_lookups.rooms') > 0) {
            resolve(ActiveLocationsAll)
        }
    })
}
