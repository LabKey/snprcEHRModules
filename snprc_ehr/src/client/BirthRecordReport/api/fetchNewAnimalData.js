import { Filter } from '@labkey/api'
import moment from 'moment'
import { request } from '../../Shared/api/api'

const parse = rows => {
    return rows.map(({ data }) => {
        return {
            Id: data.Id.value,
            value: data.Id.value,
            label: `Id: ${data.Id.value} - ${data.gender.displayValue} - Birth Date: ${moment(data.BirthDate.value).format('MM/DD/YYYY h:mm A')} `,
            BirthDate: data.BirthDate.value,
            BirthCode: data.BirthCode.value,
            AcquisitionType: data.AcquisitionType.value,
            AcqDate: data.AcqDate.value,
            gender: data.gender.value,
            sire: data.sire.value,
            dam: data.dam.value,
            species: data.species.value,
            Colony: data.Colony.value,
            AnimalAccount: data.AnimalAccount.value,
            OwnerInstitution: data.OwnerInstitution.value,
            ResponsibleInstitution: data.ResponsibleInstitution.value,
            room: data.room.value,
            Cage: data.Cage.value,
            Diet: data.Diet.value,
            pedigree: data.pedigree.value,
            IACUC: data.IACUC.value
        }
    })
}

const fetchNewAnimalData = () => {
    return new Promise((resolve, reject) => {
        request({
            schemaName: 'snprc_ehr',
            queryName: 'NewAnimalData',
            columns:
                [
                    'Id',
                    'BirthDate',
                    'BirthCode',
                    'AcquisitionType',
                    'AcqDate',
                    'gender',
                    'sire',
                    'Dam',
                    'Species',
                    'Colony',
                    'AnimalAccount',
                    'OwnerInstitution',
                    'ResponsibleInstitution',
                    'room',
                    'Cage',
                    'Diet',
                    'pedigree',
                    'IACUC'
                ],

            sort: 'Id',
            filterArray: [
                Filter.create('AcquisitionType/category', 'Birth', Filter.Types.EQUAL)
            ]
        }).then(({ rows }) => {
            resolve(parse(rows))
        }).catch(error => {
            reject(error)
            console.log('error', error)
        })
    })
}
export default fetchNewAnimalData
