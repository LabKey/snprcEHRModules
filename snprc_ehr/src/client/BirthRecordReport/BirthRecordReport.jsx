/* eslint-disable no-alert */

import React from 'react'
import { LoadingSpinner } from '@labkey/components'
import './styles/birthRecordReport.scss'
import { Button } from 'react-bootstrap'
import BirthRecordState from './constants/BirthRecordState'
import AnimalSelectionPanel from './components/AnimalSelectionPanel'
import fetchNewAnimalData from './api/fetchNewAnimalData'
import SummaryGridPanel from './components/SummaryGridPanel'
import { getReportPath } from './services/printToPDF'
import InfoPanel from '../Shared/components/InfoPanel'

export default class BirthRecordReport extends React.Component {
    state = new BirthRecordState();

    componentDidMount() {
        Promise.resolve(this.loadLists()).catch = error => {
            console.log(`Error in componentDidMount: ${error}`)
        }
    }

    loadLists() {
        let animalList = []

        return new Promise(() => {
            (
                fetchNewAnimalData().then(list => {
                    animalList = list
                })
            ).then(() => {
                this.setState(prevState => (
                    {
                        ...prevState,
                        isLoading: false,
                        animalList
                    }
                ))
            }).catch(error => {
                console.log(`Error in loadLists: ${error}`)
            })
        })
    }

    // quit
    onQuitClick = () => {
        window.history.back()
    }

    onClickPrint = () => {
        if (!this.state.selectedAnimal) {
            this.setState(prevState => (
                {
                    ...prevState,
                    errorMessage: 'Select an animal before clicking print.'
                }))
        } else {
            const reportPath = getReportPath('BirthRecord')
            const fullPath = `${reportPath}&rc:Parameters=Collapsed&TargetID=${this.state.selectedAnimal.Id}`
            const left = window.screenX + 20

            window.open(fullPath, '_blank', `location=yes,height=850,width=768,status=yes, left=${left}`)

            // add animal to the summary list if it is not there
            if (!this.state.summaryData.find(o => o.Id === this.state.selectedAnimal.Id)) {
                this.setState(prevState => (
                    {
                        ...prevState,
                        summaryData: [
                            ...prevState.summaryData,
                            { ...prevState.selectedAnimal }
                        ]
                    }))
            }
        }
    }

    handleChange = option => {
        this.setState(prevState => (
            {
                ...prevState,
                errorMessage: undefined,
                selectedAnimal: option
            }
        ))
    }

    render() {
        const { isLoading } = this.state

        if (isLoading) {
            return (
              <LoadingSpinner msg="Loading app..." />
            )
        }

        return (
          <div>
            <div className="split-panel">
              <div className="parent-panel">
                <div className="panel-heading">
                  <p>Select Animal</p>
                </div>
                <div className="wizard-panel">
                  <AnimalSelectionPanel
                    animalList={ this.state.animalList }
                    handleChange={ this.handleChange }
                    selectedAnimal={ this.state.selectedAnimal }
                  />
                  <InfoPanel
                    errorMessages={ this.state.errorMessage
                                && [{ propTest: true, colName: this.state.errorMessage }] }
                    infoMessages={ [{ key: 1, value: 'Select an animal and press the print button to generate the Birth Record Report.' }] }
                  />
                </div>
              </div>
              <div className="right-panel">
                <div className="panel-heading">
                  <p>Reports Printed</p>
                </div>
                <div className="wizard-right">
                  <SummaryGridPanel
                    summaryData={ this.state.summaryData }
                  />
                </div>
              </div>
            </div>

            <div className="wizard-footer">
              <Button bsStyle="primary" onClick={ this.onClickPrint }>Print Birth Record <i aria-hidden="true" className="fa fa-print" /></Button>
              <Button onClick={ this.onQuitClick }>Quit</Button>
            </div>
          </div>
        )
    }
}
