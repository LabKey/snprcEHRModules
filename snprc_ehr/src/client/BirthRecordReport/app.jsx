import React from 'react'
import ReactDOM from 'react-dom'

import BirthRecordReport from './BirthRecordReport'

// Need to wait for container element to be available in labkey wrapper before render
window.addEventListener('DOMContentLoaded', () => {
    ReactDOM.render(<BirthRecordReport />, document.getElementById('app'))
})
