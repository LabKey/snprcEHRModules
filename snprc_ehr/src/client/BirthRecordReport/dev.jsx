import React from 'react'
import ReactDOM from 'react-dom'
import { AppContainer } from 'react-hot-loader'

import BirthRecordReport from './BirthRecordReport'

const render = () => {
    ReactDOM.render(
      <AppContainer>
        <BirthRecordReport />
      </AppContainer>,
        document.getElementById('app')
    )
}

if (module.hot) {
    module.hot.accept()
}

render()
