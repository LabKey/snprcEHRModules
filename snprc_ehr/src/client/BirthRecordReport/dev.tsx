import React from 'react'
import ReactDOM from 'react-dom'
import { AppContainer } from 'react-hot-loader'

import BirthRecordReport from './BirthRecordReport'

const render = () => {
    ReactDOM.render(
      <AppContainer>
        <BirthRecordReport />
      </AppContainer>,
        document.getElementById('app')
    )
}

declare const module: any;

if (module.hot) {
    module.hot.accept()
}

render()
