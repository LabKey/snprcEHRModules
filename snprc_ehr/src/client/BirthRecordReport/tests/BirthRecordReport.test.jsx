/* eslint-disable no-unused-vars */

import React from 'react'
import { shallow } from 'enzyme'
import Select from 'react-select'
import BirthRecordReport from '../BirthRecordReport'
import { request } from '../../Shared/api/api'
import data from './fixtures/testData'

require("setimmediate")

jest.mock('../../Shared/api/api')

let wrapper

beforeEach(() => {
    wrapper = shallow(<BirthRecordReport />)
})

afterEach(() => {
    wrapper.unmount()
})

function flushPromises() {
    return new Promise(resolve => setImmediate(resolve))
}

describe('BirthRecordReport tests', () => {
    test('Should render a loading spinner    before api calls.', () => {
        wrapper = shallow(<BirthRecordReport />, { disableLifecycleMethods: true })
        expect(wrapper).toMatchSnapshot()
        expect(wrapper.find('LoadingSpinner').exists()).toBeTruthy()
    })

    test('Should render AnimalSelectionPanel.', async () => {
        await flushPromises()

        // spinner should be gone
        expect(wrapper.find('LoadingSpinner').exists()).toBeFalsy()

        // AnimalSelection panels should be present
        expect(wrapper.find('AnimalSelectionPanel').exists()).toBeTruthy()

        // InfoPanel should be present
        wrapper.find('AnimalSelectionPanel').dive().find('InfoPanel')

        // componentDidMount should have loaded lists for dropdowns
        expect(wrapper.state('animalList').length).toBe(2)

        // snapshot test (full page render after lists have been loaded)
        expect(wrapper).toMatchSnapshot()
    })

    test('Should render selected Animal.', async () => {
        await flushPromises()

        const animalSelectionPanel = wrapper.find('AnimalSelectionPanel')

        // Animal Select
        const animalSelect = animalSelectionPanel.dive().find(Select)

        // Test Animal Change
        animalSelect.simulate('change', data.animal)
        await flushPromises()
        expect(wrapper.state().selectedAnimal).toEqual(data.animal)

        // snapshot test (full page render after lists have been loaded)
        expect(wrapper).toMatchSnapshot()
    })
})
