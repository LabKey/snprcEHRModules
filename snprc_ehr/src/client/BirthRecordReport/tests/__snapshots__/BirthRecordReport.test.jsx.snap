// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BirthRecordReport tests Should render AnimalSelectionPanel. 1`] = `
<div>
  <div
    className="split-panel"
  >
    <div
      className="parent-panel"
    >
      <div
        className="panel-heading"
      >
        <p>
          Select Animal
        </p>
      </div>
      <div
        className="wizard-panel"
      >
        <AnimalSelectionPanel
          animalList={
            Array [
              Object {
                "AcqDate": "2020-07-08 15:04:40.000",
                "AcquisitionType": 1,
                "AnimalAccount": "0060-014-00",
                "BirthCode": 0,
                "BirthDate": "2020-07-08 15:04:40.000",
                "Cage": null,
                "Colony": 23,
                "Diet": "1",
                "IACUC": "1349MM",
                "Id": "310",
                "OwnerInstitution": 1,
                "ResponsibleInstitution": 1,
                "dam": "16355",
                "gender": "F",
                "label": "Id: 310 - Female - Birth Date: 07/08/2020 3:04 PM ",
                "pedigree": null,
                "room": 4246,
                "sire": null,
                "species": "MML",
                "value": "310",
              },
              Object {
                "AcqDate": "2020-07-09 09:13:35.000",
                "AcquisitionType": 1,
                "AnimalAccount": "0060-014-00",
                "BirthCode": 0,
                "BirthDate": "2020-07-09 09:13:35.000",
                "Cage": null,
                "Colony": 19,
                "Diet": "1",
                "IACUC": "1461CJ",
                "Id": "311",
                "OwnerInstitution": 1,
                "ResponsibleInstitution": 1,
                "dam": null,
                "gender": "F",
                "label": "Id: 311 - Female - Birth Date: 07/09/2020 9:13 AM ",
                "pedigree": null,
                "room": 4402,
                "sire": null,
                "species": "CTJ",
                "value": "311",
              },
            ]
          }
          handleChange={[Function]}
        />
        <InfoPanel
          infoMessages={
            Array [
              Object {
                "key": 1,
                "value": "Select an animal and press the print button to generate the Birth Record Report.",
              },
            ]
          }
        />
      </div>
    </div>
    <div
      className="right-panel"
    >
      <div
        className="panel-heading"
      >
        <p>
          Reports Printed
        </p>
      </div>
      <div
        className="wizard-right"
      >
        <SummaryGridPanel
          summaryData={Array []}
        />
      </div>
    </div>
  </div>
  <div
    className="wizard-footer"
  >
    <Button
      active={false}
      block={false}
      bsClass="btn"
      bsStyle="primary"
      disabled={false}
      onClick={[Function]}
    >
      Print Birth Record 
      <i
        aria-hidden="true"
        className="fa fa-print"
      />
    </Button>
    <Button
      active={false}
      block={false}
      bsClass="btn"
      bsStyle="default"
      disabled={false}
      onClick={[Function]}
    >
      Quit
    </Button>
  </div>
</div>
`;

exports[`BirthRecordReport tests Should render a loading spinner    before api calls. 1`] = `
<LoadingSpinner_LoadingSpinner
  msg="Loading app..."
  wrapperClassName=""
/>
`;

exports[`BirthRecordReport tests Should render selected Animal. 1`] = `
<div>
  <div
    className="split-panel"
  >
    <div
      className="parent-panel"
    >
      <div
        className="panel-heading"
      >
        <p>
          Select Animal
        </p>
      </div>
      <div
        className="wizard-panel"
      >
        <AnimalSelectionPanel
          animalList={
            Array [
              Object {
                "AcqDate": "2020-07-08 15:04:40.000",
                "AcquisitionType": 1,
                "AnimalAccount": "0060-014-00",
                "BirthCode": 0,
                "BirthDate": "2020-07-08 15:04:40.000",
                "Cage": null,
                "Colony": 23,
                "Diet": "1",
                "IACUC": "1349MM",
                "Id": "310",
                "OwnerInstitution": 1,
                "ResponsibleInstitution": 1,
                "dam": "16355",
                "gender": "F",
                "label": "Id: 310 - Female - Birth Date: 07/08/2020 3:04 PM ",
                "pedigree": null,
                "room": 4246,
                "sire": null,
                "species": "MML",
                "value": "310",
              },
              Object {
                "AcqDate": "2020-07-09 09:13:35.000",
                "AcquisitionType": 1,
                "AnimalAccount": "0060-014-00",
                "BirthCode": 0,
                "BirthDate": "2020-07-09 09:13:35.000",
                "Cage": null,
                "Colony": 19,
                "Diet": "1",
                "IACUC": "1461CJ",
                "Id": "311",
                "OwnerInstitution": 1,
                "ResponsibleInstitution": 1,
                "dam": null,
                "gender": "F",
                "label": "Id: 311 - Female - Birth Date: 07/09/2020 9:13 AM ",
                "pedigree": null,
                "room": 4402,
                "sire": null,
                "species": "CTJ",
                "value": "311",
              },
            ]
          }
          handleChange={[Function]}
          selectedAnimal={
            Object {
              "AcqDate": "2020-07-08 15:04:40.000",
              "AcquisitionType": 1,
              "AnimalAccount": "0060-014-00",
              "BirthCode": 0,
              "BirthDate": "2020-07-08 15:04:40.000",
              "Cage": null,
              "Colony": 23,
              "Diet": "1",
              "IACUC": "1349MM",
              "Id": "310",
              "OwnerInstitution": 1,
              "ResponsibleInstitution": 1,
              "dam": 16355,
              "gender": "F",
              "label": "Id: 310 - Female - Birth Date: 07/08/2020 3:04 PM ",
              "pedigree": null,
              "room": 4246,
              "sire": null,
              "species": "MML",
              "value": "310",
            }
          }
        />
        <InfoPanel
          infoMessages={
            Array [
              Object {
                "key": 1,
                "value": "Select an animal and press the print button to generate the Birth Record Report.",
              },
            ]
          }
        />
      </div>
    </div>
    <div
      className="right-panel"
    >
      <div
        className="panel-heading"
      >
        <p>
          Reports Printed
        </p>
      </div>
      <div
        className="wizard-right"
      >
        <SummaryGridPanel
          summaryData={Array []}
        />
      </div>
    </div>
  </div>
  <div
    className="wizard-footer"
  >
    <Button
      active={false}
      block={false}
      bsClass="btn"
      bsStyle="primary"
      disabled={false}
      onClick={[Function]}
    >
      Print Birth Record 
      <i
        aria-hidden="true"
        className="fa fa-print"
      />
    </Button>
    <Button
      active={false}
      block={false}
      bsClass="btn"
      bsStyle="default"
      disabled={false}
      onClick={[Function]}
    >
      Quit
    </Button>
  </div>
</div>
`;
